llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

```diff
@@ -451,13 +497,33 @@ bool SDWASrcOperand::convertToSDWA(MachineInstr &MI, const SIInstrInfo *TII) {
   }
   copyRegOperand(*Src, *getTargetOperand());
   if (!IsPreserveSrc) {
-    SrcSel->setImm(getSrcSel());
+    SdwaSel ExistingSel = static_cast<SdwaSel>(SrcSel->getImm());
+    SrcSel->setImm(combineSdwaSel(ExistingSel, getSrcSel()).value());
     SrcMods->setImm(getSrcMods(TII, Src));
   }
   getTargetOperand()->setIsKill(false);
   return true;
 }
 
+bool SDWASrcOperand::canCombineSelections(const MachineInstr &MI,
+                                          const SIInstrInfo *TII) {
+  if (!TII->isSDWA(MI.getOpcode()))
+    return true;
+
+  auto canCombineSel = [&](auto SrcOpName, auto SrcSelOpName) {
```

Comment:
Move this to a full helper instead of the lambda?

Reply:
Done.